mava/networks/sable_network.py

```diff
@@ -0,0 +1,547 @@
+        """Apply chunkwise decoding."""
+        # Initialize the updated hidden states
+        updated_hstates = (jnp.zeros_like(hstates[0]), jnp.zeros_like(hstates[1]))
+        # Encode the action
+        action_embeddings = self.action_encoder(action)
+        x = self.ln(action_embeddings)
+
+        # Apply the decoder blocks
+        for i, block in enumerate(self.blocks):
+            hs = jax.tree.map(lambda x, j=i: x[:, :, j], hstates)
+            x, hs_new = block(
+                x=x, obs_rep=obs_rep, hstates=hs, dones=dones, timestep_id=timestep_id
+            )
+            updated_hstates = jax.tree.map(
+                lambda x, y, j=i: x.at[:, :, j].set(y), updated_hstates, hs_new
+            )
+
+        # Compute the action logits
+        logit = self.head(x)
+
+        return logit, updated_hstates
+
+    def recurrent(
+        self,
+        action: chex.Array,
+        obs_rep: chex.Array,
+        hstates: Tuple[chex.Array, chex.Array],
+        timestep_id: chex.Array,
+    ) -> Tuple[chex.Array, Tuple[chex.Array, chex.Array]]:
+        """Apply recurrent decoding."""
+        # Initialize the updated hidden states
+        updated_hstates = (jnp.zeros_like(hstates[0]), jnp.zeros_like(hstates[1]))
+        # Encode the action
+        action_embeddings = self.action_encoder(action)
+        x = self.ln(action_embeddings)
+
+        # Apply the decoder blocks
+        for i, block in enumerate(self.blocks):
+            hs = jax.tree.map(lambda x, i=i: x[:, :, i], hstates)
+            x, hs_new = block.recurrent(x=x, obs_rep=obs_rep, hstates=hs, timestep_id=timestep_id)
+            updated_hstates = jax.tree.map(
+                lambda x, y, j=i: x.at[:, :, j].set(y), updated_hstates, hs_new
+            )
+
+        # Compute the action logits
+        logit = self.head(x)
+
+        return logit, updated_hstates
+
+
+class SableNetwork(nn.Module):
+    """Sable network module."""
+
+    n_agents: int
+    action_dim: int
+    net_config: SableNetworkConfig
+    memory_config: DictConfig
+    action_space_type: str = "discrete"
+
+    def setup(self) -> None:
+        # Check if the action space type is valid
+        assert self.action_space_type in [
+            "discrete",
+            "continuous",
+        ], "Invalid action space type"
+
+        self.n_agents_per_chunk = self.n_agents
+        if self.memory_config.use_chunkwise:
+            if self.memory_config.type == "ff_sable":
+                self.memory_config.chunk_size = self.memory_config.agents_chunk_size
+                assert (
+                    self.n_agents % self.memory_config.chunk_size == 0
+                ), "Number of agents should be divisible by chunk size"
+                self.n_agents_per_chunk = self.memory_config.chunk_size
+            else:
+                self.memory_config.chunk_size = (
+                    self.memory_config.timestep_chunk_size * self.n_agents
+                )
```

Comment:
I am not sure how it will be in system instead of network, it will be too complicated however at least what we have now the config that is logged to neptune is correct what happen inside here just unification of the params that we use in memory vs non memory